### PR TITLE
Windows native toasts

### PR DESCRIPTION
**Pull request by alameenshah**
#### Supported notifications
| Notification | Screenshot | Action |
| --- | --- |---|
| Autmount start |Automount_Start|This notification is not actionable |
| Mount failure during auto-mount |Automount_Failure|Automount_Failure_Response|
| Upgrade available | Upgrade_Available| Upgrade_Available_Response|

#### Changes in this PR
- `VFSForGit` installer creates a `Start Menu` shortcut to `GVFS.Service.UI.exe` after successful install. (This shortcut is a requirement for enabling native toasters.)
- `VFSForGit` installer launches `GVFS.Service.UI.exe` in the currently logged-in user context after install
- `GVFS.Service` will launch `GVFS.Service.UI` when a new user logs in.
- `GVFS.Service.UI.exe` will always be running.
- `GVFS.Service.UI.exe` listens for supported notifications. Notifications that cannot be identified are ignored.
- Auto-mount and Upgrade available toasters are actionable. In response these actionable toasters, `GVFS.Service.UI` runs `gvfs mount` and `gvfs upgrade --confirm` (elevated) verbs in a new Command Prompt window.
- `GVFS.Service` waits for a minute before doing upgrade checks, just so that `VFSForGit` installer gets the time to launch `GVFS.Service.UI` that is needed to display the notifications.

#### TODO
- [x] Testing
- [x] Integrate with `GVFS.Service`
- [x] Unit testing
- [x] Update diagnose verb to cat `GVFS.Service.UI` logs

### CODE REVIEW

File: GVFS/GVFS.Installer.Windows/Setup.iss [hunk 1]
```diff
@@ -27,6 +27,7 @@
 #define EnvironmentKey "SYSTEM\CurrentControlSet\Control\Session Manager\Environment"
 #define FileSystemKey "SYSTEM\CurrentControlSet\Control\FileSystem"
 #define GvFltAutologgerKey "SYSTEM\CurrentControlSet\Control\WMI\Autologger\Microsoft-Windows-Git-Filter-Log"
+#define ServiceUIName "GVFS.Service.UI"
```
Comment (mjcheetham) [inline GVFS/GVFS.Installer.Windows/Setup.iss]: Do we want to give the UI a nicer name (assuming the notification sender is based on the shortcut name), like "VFS for Git"?

image


```suggestion
#define ServiceUIName "VFS for Git"
```

File: GVFS/GVFS.Service.UI/GVFS.Service.UI.csproj [hunk 2]
```diff
@@ -86,5 +91,6 @@
     <Analyzer Include="..\..\..\packages\StyleCop.Analyzers.1.0.2\analyzers\dotnet\cs\StyleCop.Analyzers.CodeFixes.dll" />
     <Analyzer Include="..\..\..\packages\StyleCop.Analyzers.1.0.2\analyzers\dotnet\cs\StyleCop.Analyzers.dll" />
   </ItemGroup>
+  <ItemGroup />
```
Comment (mjcheetham) [inline GVFS/GVFS.Service.UI/GVFS.Service.UI.csproj]: nit: Remove the empty `<ItemGroup/>` please 😄 

```suggestion
```

File: GVFS/GVFS.Service.UI/GVFSToastRequestHandler.cs [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+using GVFS.Common.NamedPipes;
+using GVFS.Common.Tracing;
+using System;
+using System.Diagnostics;
+
+namespace GVFS.Service.UI
+{
+    public class GVFSToastRequestHandler
+    {
+        private const string GVFSAutomountStartTitle= "GVFS Automount";
```
Comment (mjcheetham) [inline GVFS/GVFS.Service.UI/GVFSToastRequestHandler.cs]: Shouldn't these be named "VFS ..."?

```suggestion
        private const string GVFSAutomountStartTitle= "VFS Automount";
```

Comment (mjcheetham) [reply GVFS/GVFS.Service.UI/GVFSToastRequestHandler.cs]: Same applies to the rest of the strings here (and the property names?).

File: GVFS/GVFS.Service.UI/GVFSToastRequestHandler.cs [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        private const string GVFSAutomountButtonTitle = "Retry";
+
+        private const string GVFSUpgradeTitleFormat = "New version {0} is available";
+        private const string GVFSUpgradeMessage = "Upgrade will unmount and remount gvfs repos, ensure you are at a stopping point. When ready, click Upgrade button to run upgrade.";
+        private const string GVFSUpgradeButtonTitle = "Upgrade";
+
+        private const string GvfsRemountActionPrefix = "gvfs mount ";
+        private const string GvfsUpgradeActionPrefix = "gvfs upgrade --confirm";
+
+        private ITracer tracer;
```
Comment (mjcheetham) [inline GVFS/GVFS.Service.UI/GVFSToastRequestHandler.cs]: Can make these fields `readonly`.

File: GVFS/GVFS.Service.UI/GVFSToastRequestHandler.cs [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+                        message = string.Format(GVFSAutomountSuccessMessageFormat, Environment.NewLine, request.Enlistment);
+                        break;
+                    }
+
+                case NamedPipeMessages.Notification.Request.Identifier.MountFailure:
+                    {
+                        title = GVFSAutomountErrorTitle;
+                        message = string.Format(GVFSAutomountErrorMessageFormat, Environment.NewLine, request.Enlistment);
+                        buttonTitle = GVFSAutomountButtonTitle;
+                        args = $"{GvfsRemountActionPrefix} {request.Enlistment}";
```
Comment (mjcheetham) [inline GVFS/GVFS.Service.UI/GVFSToastRequestHandler.cs]: Do we have to worry about escaping the enlistment here (spaces in the path) or is that done elsewhere?

File: GVFS/GVFS.Service.UI/GVFSToastRequestHandler.cs [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+        public void UserResponseCallback(string args)
+        {
+            if (!string.IsNullOrEmpty(args))
+            {
+                string gvfsArgs = null;
+                bool elevate = false;
+
+                if (GVFSPlatform.Instance.IsElevated())
+                {
+                    this.tracer.RelatedInfo($"GVFS.Serice.UI is currently running elevated.");
```
Comment (mjcheetham) [inline GVFS/GVFS.Service.UI/GVFSToastRequestHandler.cs]: Does this mean anyone can send a toast notification to `GVFS.Service.UI.exe` and have their code run elevated? I know we only accept some predefined 'actions' for the list of supported notification types, but if there are ways of breaking out, we should be careful here!

This is why I ask about escaping the 'enlistment path' args that we get back from a notification callback, if we're just shelling out to a possibly elevated cmd.exe.

File: GVFS/GVFS.Installer.Windows/Setup.iss (hunk 1 above)
Comment (alameenshah) [reply GVFS/GVFS.Installer.Windows/Setup.iss]: Updated it to `VFS For Git`

File: GVFS/GVFS.Service.UI/GVFSToastRequestHandler.cs (hunk 3 above)
Comment (alameenshah) [reply GVFS/GVFS.Service.UI/GVFSToastRequestHandler.cs]: Updated the message text and properties in this file. (But there are so many properties, classes and files that use gvfs in their names today. )

File: GVFS/GVFS.Service.UI/GVFSToastRequestHandler.cs (hunk 5 above)
Comment (alameenshah) [reply GVFS/GVFS.Service.UI/GVFSToastRequestHandler.cs]: Added path validation both while sending and while receiving callbacks from toaster.

File: GVFS/GVFS.Service.UI/GVFSToastRequestHandler.cs (hunk 6 above)
Comment (alameenshah) [reply GVFS/GVFS.Service.UI/GVFSToastRequestHandler.cs]: >>Does this mean anyone can send a toast notification to GVFS.Service.UI.exe and have their code run elevated?

- `GVFS.Service.UI` is running un-elevated. So it cannot. (However it runs `gvfs upgrade --confirm` command elevated after user approval).
- Also added path validation.

File: GVFS/GVFS/CommandLine/DiagnoseVerb.cs [hunk 7]
```diff
@@ -33,6 +34,8 @@ protected override string VerbName
 
         protected override void Execute(GVFSEnlistment enlistment)
         {
+            Debugger.Launch();
```
Comment (kewillford) [inline GVFS/GVFS/CommandLine/DiagnoseVerb.cs]: Do we still need this?

File: GVFS/GVFS.Service/GVFSService.Windows.cs [hunk 8]
```diff
@@ -365,5 +370,40 @@ private DirectorySecurity GetServiceDirectorySecurity(string serviceDataRootPath
...
         }
+
+        private void LaunchServiceUIIfNotRunning(int sessionId)
+        {
+            NamedPipeClient client;
+            using (client = new NamedPipeClient(GVFSConstants.Service.UIName))
+            {
+                if (!client.Connect())
+                {
+                    this.TerminateExistingProcess(GVFSConstants.Service.UIName);
```
Comment (kewillford) [inline GVFS/GVFS.Service/GVFSService.Windows.cs]: Help me understand why we are killing the process and restarting it if we can't connect?

File: GVFS/GVFS.Service/GVFSService.Windows.cs [hunk 9]
```diff
@@ -320,23 +326,22 @@ private void CreateAndConfigureProgramDataDirectories()
             Directory.SetAccessControl(serviceDataRootPath, serviceDataRootSecurity);
 
             // Special rules for the upgrader logs, as non-elevated users need to be be able to write
-            this.CreateAndConfigureUpgradeLogDirectory();
+            this.CreateAndConfigureLogDirectory(ProductUpgraderInfo.GetLogDirectoryPath());
+            this.CreateAndConfigureLogDirectory(GVFSPlatform.Instance.GetDataRootForGVFSComponent(GVFSConstants.Service.UIName));
         }
 
-        private void CreateAndConfigureUpgradeLogDirectory()
+        private void CreateAndConfigureLogDirectory(string path)
```
Comment (kewillford) [inline GVFS/GVFS.Service/GVFSService.Windows.cs]: Do we need to pass something to tell us what path this is?  Either the upgrader or the service or will the name of the log be enough for us to know which one?

File: GVFS/GVFS.Service.UI/WinToastNotifier.cs [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+
+            ToastNotifier toastNotifier = ToastNotificationManager.CreateToastNotifier(this.appId);
+            toastNotifier.Show(toastNotification);
+        }
+
+        private void ToastActivated(ToastNotification sender, object e)
+        {
+            ToastActivatedEventArgs args = (ToastActivatedEventArgs)e;
+
+            this.UserResponseCallback?.Invoke(args.Arguments);
```
Comment (kewillford) [inline GVFS/GVFS.Service.UI/WinToastNotifier.cs]: Do we need a `try/catch` around this or in the callback method?

File: GVFS/GVFS.Service/GVFSService.Windows.cs (hunk 8 above)
Comment (alameenshah) [reply GVFS/GVFS.Service/GVFSService.Windows.cs]: This is old code that is reused ([here is the original](https://github.com/microsoft/VFSForGit/blob/686b3b1f82021b515bf1d4dce728b9b90925a13a/GVFS/GVFS.Service/Handlers/NotificationHandler.Windows.cs#L19) in master.)

And probably is not needed. @wilbaker thoughts?

File: GVFS/GVFS.Service.UI/GVFSToastRequestHandler.cs [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+            string message = null;
+            string buttonTitle = null;
+            string args = null;
+            string path = null;
+
+            switch (request.Id)
+            {
+                case NamedPipeMessages.Notification.Request.Identifier.AutomountStart:
+                    title = GVFSAutomountStartTitle;
+                    message = string.Format(GVFSAutomountStartMessageFormat, request.EnlistmentCount);
```
[the remark this answers is not in the document]
Comment (alameenshah) [reply GVFS/GVFS.Service.UI/GVFSToastRequestHandler.cs]: I'll then have to call it for each `switch` `case`.  This pattern (customize the local variables for each `switch case` and call `Notify()` just once in the end) is simpler.

File: GVFS/GVFS/CommandLine/DiagnoseVerb.cs (hunk 7 above)
Comment (alameenshah) [reply GVFS/GVFS/CommandLine/DiagnoseVerb.cs]: Nope. Removed.

File: GVFS/GVFS.Service/GVFSService.Windows.cs (hunk 9 above)
Comment (alameenshah) [reply GVFS/GVFS.Service/GVFSService.Windows.cs]: `CreateAndConfigureLogDirectory` is a generic method. It doesn't need to know what process/function the log-directory is associated with.

File: GVFS/GVFS.Service.UI/WinToastNotifier.cs (hunk 10 above)
Comment (alameenshah) [reply GVFS/GVFS.Service.UI/WinToastNotifier.cs]: Yes... I've added it around the `Process.Start` inside the callback. The other place that could throw exception is `TryValidatePath` which already has a `catch` block.

File: GVFS/GVFS.Service/GVFSService.Windows.cs (hunk 8 above)
Comment (wilbaker) [reply GVFS/GVFS.Service/GVFSService.Windows.cs]: > And probably is not needed. @wilbaker thoughts?

I'm not familiar with the service UI code.  Have you tried going back in history (before the move to GitHub) to see if you can find any discussion on why the code is structured this way?

File: GVFS/GVFS.Service.UI/GVFSToastRequestHandler.cs [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+        {
+            const string cmdPath = "CMD.exe";
+            ProcessStartInfo processInfo = new ProcessStartInfo(cmdPath);
+            processInfo.UseShellExecute = true;
+            processInfo.RedirectStandardInput = false;
+            processInfo.RedirectStandardOutput = false;
+            processInfo.RedirectStandardError = false;
+            processInfo.WindowStyle = ProcessWindowStyle.Normal;
+            processInfo.CreateNoWindow = false;
+            processInfo.Arguments = "/K " + fullGvfsCmd;
```
[the remark this answers is not in the document]
Comment (alameenshah) [reply GVFS/GVFS.Service.UI/GVFSToastRequestHandler.cs]: Yes. This is especially useful during an error, user gets the time to read it on the `Command` window and then close it manually.

I will add a comment in the code as well.

**Comment by alameenshah** [comment]
> Just some simple questions/comments. I was also unable to get the toasts to work. Do I need to do anything besides installing using the GVFS installer? I only had 1 repo that was automounted so maybe it happened too fast?

No, just installation is enought. We don't show notifications on Mount Failures (Not success) and when an upgrade is available. Can you simulate a mount failure, or an upgrade available situation and see.

File: GVFS/GVFS.Installer.Windows/Setup.iss [hunk 13]
```diff
@@ -769,6 +803,7 @@ begin
       Abort();
     end;
   StopService('GVFS.Service');
+  StopGVFSServiceUI();
```
[the remark this answers is not in the document]
Comment (alameenshah) [reply GVFS/GVFS.Installer.Windows/Setup.iss]: `GVFS.Service.UI` is not a service. It is launched as hidden process in the current users sessionId.

File: GVFS/GVFS.Service.UI/Program.cs [hunk 14]
```diff
@@ -2,15 +2,9 @@
 using GVFS.Common.NamedPipes;
```
[the remark this answers is not in the document]
Comment (alameenshah) [reply GVFS/GVFS.Service.UI/Program.cs]: Moved `GVFS.Service.UI` to a separate file. `Program.cs` now only has the `Main` entry.

**Pull request merged**